mitiq/folding_cirq.py

```diff
@@ -0,0 +1,97 @@
+from typing import List, Callable
+import random
+from cirq import Circuit, inverse
+
+
+# ==================================================
+# Gate level folding
+# ==================================================
+
+
+def sampling_stretcher(circuit: Circuit, stretch: float) -> Circuit:
+    """Applies the map G -> G G^dag G to a random subset of gates 
+    of the input circuit.
+    Returns a circuit of depth approximately equal to stretch*len(circuit).
+    The stretch factor can be any real number within 1 and 3."""
+
+    if not ((stretch >= 1) and (stretch <= 3)):
+        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+
+    depth = len(circuit)
+    reduced_depth = int(depth * (stretch - 1) / 2)
+    sub_indices = random.sample(range(depth), reduced_depth)
+    return fold_gates(circuit, sub_indices)
+
+
+def left_stretcher(circuit: Circuit, stretch: float) -> Circuit:
+    """Applies the map G -> G G^dag G to a subset of gates of the input circuit (first part of it).
+    Returns a circuit of depth approximately equal to stretch*len(circuit).
+    The stretch factor can be any real number within 1 and 3."""
+
+    if not ((stretch >= 1) and (stretch <= 3)):
+        raise ValueError("The stretch factor must be a real number within 1 and 3.")
+
+    depth = len(circuit)
+    reduced_depth = int(depth * (stretch - 1) / 2)
+    sub_indices = list(range(reduced_depth))
+    return fold_gates(circuit, sub_indices)
+
+
+def fold_gates(circuit: Circuit, sub_indices: List[int]) -> Circuit:
+    """Applies the map G -> G G^dag G to a subset of gates of the input circuit
+    determined by sub_indices."""
+
+    out = Circuit()
+    for j, gate in enumerate(circuit):
```

Comment:
I don't think the behavior of this function is what is intended.

When you iterate over a cirq.Circuit, you iterate over moments, which are sets of gates which happen in the same time slice. You can still apply `cirq.inverse` to a moment, but due to the name `gate`, I don't think this is the intended behavior of the function.

The test cases in [](mitiq/tests/test_folding_cirq.py) only test single qubit circuits, for which the behavior is identical, but I think this function has un-intended consequences for multi-qubit circuits.

Reply:
@rmlarose Great point. This follows on from the discussion that we had the other day on the phone. Can you please open up an issue? Please do suggest a new PR/fix into master if you have one in mind.

Reply:
My bad for missing this in code review. But luckily we have two sets of eyes!